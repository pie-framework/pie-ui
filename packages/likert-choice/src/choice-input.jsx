import {FormControlLabel} from 'material-ui/Form';
import React from 'react';
import PropTypes from 'prop-types';
import {withStyles} from 'material-ui/styles';
import Radio from 'material-ui/Radio';
import green from 'material-ui/colors/green';
import orange from 'material-ui/colors/orange';
import Typography from 'material-ui/typography';
import classNames from 'classnames';

const styleSheet = {
    title: {
        fontWeight: 'bold',
        fontSize: '1em',
        paddingRight: '1em'
    },
    label: {
        color: 'var(--choice-input-color, black)',
        display: 'block',
        verticalAlign: 'middle',
        cursor: 'pointer'
    },
    row: {
        display: 'flex',
        alignItems: 'center'
    }
};

const formStyleSheet = {
    label: {
        color: 'var(--choice-input-color, black)'
    }
};

export const StyledFormControlLabel = withStyles(formStyleSheet, {
    name: 'FormControlLabel'
})(props => (
    <FormControlLabel {...props} classes={{label: props.classes.label}}/>
));

const CLASS_NAME = 'multiple-choice-component';

const colorStyle = (varName, fallback) => ({
    [`&.${CLASS_NAME}`]: {
        color: `var(--choice-input-${varName}, ${fallback})`
    }
});

const inputStyles = {
    'correct-root': colorStyle('correct-color', 'black'),
    'correct-checked': colorStyle('correct-selected-color', green[500]),
    'correct-disabled': colorStyle('correct-disabled-color', 'grey'),
    'incorrect-root': colorStyle('incorrect-color', 'black'),
    'incorrect-checked': colorStyle('incorrect-checked', orange[500]),
    'incorrect-disabled': colorStyle('incorrect-disabled-color', 'grey'),
    root: colorStyle('color', 'black'),
    checked: colorStyle('selected-color', 'black'),
    disabled: colorStyle('disabled-color', 'black')
};

export const StyledRadio = withStyles(inputStyles)(props => {
    const {correctness, classes, checked, onChange, disabled} = props;
    const key = k => (correctness ? `${correctness}-${k}` : k);

    const resolved = {
        root: classes[key('root')],
        checked: classes[key('checked')],
        disabled: classes[key('disabled')]
    };

    const miniProps = {checked, onChange, disabled};

    return (
        <Radio
            {...miniProps}
            className={CLASS_NAME}
            classes={{
                root: resolved.root,
                checked: resolved.checked
            }}
        />
    );
});

export class ChoiceInput extends React.Component {
    static propTypes = {
        choiceMode: PropTypes.oneOf(['radio']),
        displayKey: PropTypes.string.isRequired,
        checked: PropTypes.bool.isRequired,
        correctness: PropTypes.string,
        disabled: PropTypes.bool.isRequired,
        label: PropTypes.string.isRequired,
        onChange: PropTypes.func.isRequired,
        value: PropTypes.string.isRequired,
        classes: PropTypes.object,
        className: PropTypes.string
    };

    static defaultProps = {};

    constructor(props) {
        super(props);
        this.onToggleChoice = this.onToggleChoice.bind(this);
    }

    onToggleChoice() {
        this.props.onChange({
            value: this.props.value,
            selected: !this.props.checked
        });
    }

    render() {
        const {
            choiceMode,
            disabled,
            displayKey,
            label,
            checked,
            correctness,
            classes,
            className
        } = this.props;

        const Tag = StyledRadio;
        const classSuffix = 'radio-button';
        return (
            <div className={classNames(className, 'corespring-' + classSuffix)}>
                <div className={classes.row}>
                    <Typography className={classes.title}>{displayKey + '. '}</Typography>
                    <StyledFormControlLabel
                        disabled={disabled}
                        control={
                            <Tag
                                checked={checked}
                                correctness={correctness}
                                onChange={this.onToggleChoice}
                            />
                        }
                    />
                    <span
                        className={classes.label}
                        onClick={this.onToggleChoice}
                        dangerouslySetInnerHTML={{__html: label}}
                    />
                </div>
            </div>
        );
    }
}

export default withStyles(styleSheet)(ChoiceInput);