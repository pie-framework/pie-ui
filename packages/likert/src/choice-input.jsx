import React from 'react';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import PropTypes from 'prop-types';
import {withStyles} from '@material-ui/core/styles';

import {color} from '@pie-lib/render-ui';
import Radio from '@material-ui/core/Radio';
import {LIKERT_ORIENTATION} from './likertEntities';

const radioStyles = {
  root: {
    color: `var(--choice-input-color, ${color.text()})`
  },
  checked: {
    color: `var(--choice-input-selected-color, ${color.text()})`
  },
};

export const RadioStyled = withStyles(radioStyles)((props) => {
  const {classes, checked, onChange, disabled} = props;

  return (
    <Radio
      checked={checked}
      onChange={onChange}
      disabled={disabled}
      classes={{
        root: classes.root,
        checked: classes.checked,
      }}
    />
  );
});


const choiceInputStyles = () => ({
  labelRoot: {
    color: color.text(),
    textAlign: 'center',
    cursor: 'pointer'
  },
  checkboxHolderRoot: {
    display: 'flex',
    alignItems: 'center',
    flex: 1,
    padding: '0 5px',
    '& label': {}
  },
  formControlLabelRoot: {
    margin: 0
  }
});


export class ChoiceInput extends React.Component {
  static propTypes = {
    checked: PropTypes.bool.isRequired,
    disabled: PropTypes.bool.isRequired,
    label: PropTypes.string.isRequired,
    likertOrientation: PropTypes.string.isRequired,
    onChange: PropTypes.func.isRequired,
    value: PropTypes.number.isRequired,
    classes: PropTypes.object,
  };

  static defaultProps = {
    checked: false,
  };

  onToggleChoice = () => {
    this.props.onChange({
      value: this.props.value,
      selected: !this.props.checked,
    });
  };

  render() {
    const {
      disabled,
      label,
      checked,
      likertOrientation,
      classes
    } = this.props;
    const flexDirection = likertOrientation === LIKERT_ORIENTATION.vertical ? 'row': 'column' ;

    return (
      <div className={classes.checkboxHolderRoot} style={{flexDirection}}>
        <FormControlLabel
          disabled={disabled}
          className={classes.formControlLabelRoot}
          control={
            <RadioStyled
              checked={checked}
              onChange={this.onToggleChoice}
            />
          }
        />
        <p
          className={classes.labelRoot}
          onClick={this.onToggleChoice}
          dangerouslySetInnerHTML={{__html: label}}
        />
      </div>
    );
  }
}

export default withStyles(choiceInputStyles)(ChoiceInput);
