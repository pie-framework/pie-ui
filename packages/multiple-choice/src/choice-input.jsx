import FormControlLabel from '@material-ui/core/FormControlLabel';
import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';

import Checkbox from '@material-ui/core/Checkbox';
import { Feedback, color } from '@pie-lib/render-ui';
import FeedbackTick from './feedback-tick';
import Radio from '@material-ui/core/Radio';
import classNames from 'classnames';

const styleSheet = (theme) => ({
  label: {
    color: `${color.text()} !important`, //'var(--choice-input-color, black)',
    display: 'inline-block',
    verticalAlign: 'middle',
    cursor: 'pointer',
  },
  row: {
    display: 'flex',
    alignItems: 'center',
    backgroundColor: color.background(),
  },
  checkboxHolder: {
    display: 'flex',
    alignItems: 'center',
    backgroundColor: color.background(),
    flex: 1,
    '& label': {
      color: color.text(),
    },
  },
  rationale: {
    paddingLeft: theme.spacing.unit * 16,
  },
});

const formStyleSheet = {
  label: {
    color: `${color.text()} !important`, //'var(--choice-input-color, black)'
    backgroundColor: color.background(),
  },
};

export const StyledFormControlLabel = withStyles(formStyleSheet, {
  name: 'FormControlLabel',
})((props) => (
  <FormControlLabel {...props} classes={{ label: props.classes.label }} />
));

const CLASS_NAME = 'multiple-choice-component';

const colorStyle = (varName, fallback) => ({
  [`&.${CLASS_NAME}`]: {
    color: `var(--choice-input-${varName}, ${fallback}) !important`,
  },
});

const inputStyles = {
  'correct-root': colorStyle('correct-color', color.text()),
  'correct-checked': colorStyle('correct-selected-color', color.correct()), //green[500]),
  'correct-disabled': colorStyle('correct-disabled-color', color.disabled()), //'grey'),
  'incorrect-root': colorStyle('incorrect-color', color.incorrect()),
  'incorrect-checked': colorStyle('incorrect-checked', color.incorrect()), //orange[500]),
  'incorrect-disabled': colorStyle(
    'incorrect-disabled-color',
    color.disabled()
  ),
  root: {
    ...colorStyle('color', color.text()),
    '&:hover': { color: `${color.primaryLight()} !important` },
  },
  checked: colorStyle('selected-color', color.primary()),
  disabled: {
    ...colorStyle('disabled-color', color.text()),
    opacity: 0.6,
    cursor: 'not-allowed !important',
    pointerEvents: 'initial !important',
  },
};

export const StyledCheckbox = withStyles(inputStyles)((props) => {
  const { correctness, classes, checked, onChange, disabled } = props;
  const key = (k) => (correctness ? `${correctness}-${k}` : k);

  const resolved = {
    root: classes[key('root')],
    checked: classes[key('checked')],
    disabled: classes[key('disabled')],
  };

  const miniProps = { checked, onChange, disabled };
  return (
    <Checkbox
      {...miniProps}
      className={CLASS_NAME}
      classes={{
        root: resolved.root,
        checked: resolved.checked,
        disabled: resolved.disabled,
      }}
    />
  );
});

export const StyledRadio = withStyles(inputStyles)((props) => {
  const { correctness, classes, checked, onChange, disabled } = props;
  const key = (k) => (correctness ? `${correctness}-${k}` : k);

  const resolved = {
    root: classes[key('root')],
    checked: classes[key('checked')],
    disabled: classes[key('disabled')],
  };

  const miniProps = { checked, onChange, disabled };

  return (
    <Radio
      {...miniProps}
      className={CLASS_NAME}
      classes={{
        root: resolved.root,
        checked: resolved.checked,
      }}
    />
  );
});

export class ChoiceInput extends React.Component {
  static propTypes = {
    choiceMode: PropTypes.oneOf(['radio', 'checkbox']),
    displayKey: PropTypes.string.isRequired,
    checked: PropTypes.bool.isRequired,
    correctness: PropTypes.string,
    disabled: PropTypes.bool.isRequired,
    feedback: PropTypes.string,
    label: PropTypes.string.isRequired,
    rationale: PropTypes.string,
    onChange: PropTypes.func.isRequired,
    value: PropTypes.string.isRequired,
    classes: PropTypes.object,
    className: PropTypes.string,
  };

  static defaultProps = {
    rationale: null,
    checked: false,
  };

  constructor(props) {
    super(props);
    this.onToggleChoice = this.onToggleChoice.bind(this);
  }

  onToggleChoice() {
    this.props.onChange({
      value: this.props.value,
      selected: !this.props.checked,
    });
  }

  render() {
    const {
      choiceMode,
      disabled,
      displayKey,
      feedback,
      label,
      checked,
      correctness,
      classes,
      className,
      rationale,
    } = this.props;

    const Tag = choiceMode === 'checkbox' ? StyledCheckbox : StyledRadio;
    const classSuffix = choiceMode === 'checkbox' ? 'checkbox' : 'radio-button';

    return (
      <div className={classNames(className, 'corespring-' + classSuffix)}>
        <div className={classes.row}>
          <FeedbackTick correctness={correctness} />
          <div className={classes.checkboxHolder}>
            <StyledFormControlLabel
              disabled={disabled}
              label={displayKey ? displayKey + '. ' : ''}
              control={
                <Tag
                  checked={checked}
                  correctness={correctness}
                  onChange={this.onToggleChoice}
                />
              }
            />
            <span
              className={classes.label}
              onClick={this.onToggleChoice}
              dangerouslySetInnerHTML={{ __html: label }}
            />
          </div>
        </div>
        {rationale && (
          <div
            className={classes.rationale}
            dangerouslySetInnerHTML={{ __html: rationale }}
          />
        )}
        <Feedback feedback={feedback} correctness={correctness} />
      </div>
    );
  }
}

export default withStyles(styleSheet)(ChoiceInput);
