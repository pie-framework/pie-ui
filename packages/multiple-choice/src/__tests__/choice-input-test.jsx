import React from 'react';
import _ from 'lodash';
import { shallow, mount } from 'enzyme';
import { ChoiceInput } from '../choice-input';
import toJson from 'enzyme-to-json';

describe('ChoiceInput', () => {
  let onChange, wrapper;

  const mkWrapper = (opts = {}) => {
    opts = _.extend(
      {
        checked: false,
        disabled: false,
        choiceMode: 'checkbox',
        label: 'label',
        displayKey: '1',
        correctness: 'correct',
        value: 'value',
        classes: {
          label: 'label'
        }
      },
      opts
    );

    return shallow(<ChoiceInput {...opts} onChange={onChange} />);
  };

  beforeEach(() => {
    onChange = jest.fn();
    wrapper = mkWrapper();
  });

  describe('snapshots', () => {
    describe('radio', () => {
      it('renders', () => {
        const wrapper = mkWrapper({ choiceMode: 'radio' });
        expect(toJson(wrapper)).toMatchSnapshot();
      });
    });

    describe('checkbox', () => {
      it('renders', () => {
        const wrapper = mkWrapper({ choiceMode: 'checkbox' });
        expect(toJson(wrapper)).toMatchSnapshot();
      });
    });

    describe('radio with incorrect', () => {
      it('renders', () => {
        const wrapper = mkWrapper({ choiceMode: 'radio', correctness: 'incorrect' });
        expect(toJson(wrapper)).toMatchSnapshot();
      });
    });

    describe('rationale', () => {
      it('does not render', () => {
        const wrapper = mkWrapper();
        expect(toJson(wrapper)).toMatchSnapshot();
      });

      it('renders', () => {
        const wrapper = mkWrapper({ rationale: 'This is rationale' });
        expect(toJson(wrapper)).toMatchSnapshot();
      });
    });

  });

  describe('onToggleChoice', () => {
    it('calls handler', () => {
      wrapper.instance().onToggleChoice({ target: { checked: true } });
      expect(onChange).toBeCalledWith({
        value: 'value',
        selected: true
      });
    });
  });

  describe('clicking on choice text triggers toggle choice', () => {
    it('calls handler', () => {
      const onChange = jest.fn();
      const props =       {
        checked: false,
        disabled: false,
        choiceMode: 'checkbox',
        label: 'label',
        displayKey: '1',
        correctness: 'correct',
        value: 'value',
        classes: {
          label: 'label'
        },
        onChange
      };
      const item = mount(<ChoiceInput { ...props } />);
      const label = item.find('.label');

      label.simulate('click');

      expect(onChange).toHaveBeenCalledWith({ value: 'value', selected: true });
    });
  });
});
