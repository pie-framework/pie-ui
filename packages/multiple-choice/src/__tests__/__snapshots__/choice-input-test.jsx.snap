// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChoiceInput snapshots checkbox renders 1`] = `
<div
  className="corespring-checkbox"
>
  <div>
    <WithStyles(FeedbackTick)
      correctness="correct"
    />
    <div>
      <WithStyles(Component)
        control={
          <WithStyles(Component)
            checked={false}
            correctness="correct"
            onChange={[Function]}
          />
        }
        disabled={false}
        label="1. "
      />
      <Component
        className="label"
        onClick={[Function]}
        prompt="label"
        tagName="span"
      />
    </div>
  </div>
  <Feedback
    correctness="correct"
  />
</div>
`;

exports[`ChoiceInput snapshots radio renders 1`] = `
<div
  className="corespring-radio-button"
>
  <div>
    <WithStyles(FeedbackTick)
      correctness="correct"
    />
    <div>
      <WithStyles(Component)
        control={
          <WithStyles(Component)
            checked={false}
            correctness="correct"
            onChange={[Function]}
          />
        }
        disabled={false}
        label="1. "
      />
      <Component
        className="label"
        onClick={[Function]}
        prompt="label"
        tagName="span"
      />
    </div>
  </div>
  <Feedback
    correctness="correct"
  />
</div>
`;

exports[`ChoiceInput snapshots radio with incorrect renders 1`] = `
<div
  className="corespring-radio-button"
>
  <div>
    <WithStyles(FeedbackTick)
      correctness="incorrect"
    />
    <div>
      <WithStyles(Component)
        control={
          <WithStyles(Component)
            checked={false}
            correctness="incorrect"
            onChange={[Function]}
          />
        }
        disabled={false}
        label="1. "
      />
      <Component
        className="label"
        onClick={[Function]}
        prompt="label"
        tagName="span"
      />
    </div>
  </div>
  <Feedback
    correctness="incorrect"
  />
</div>
`;

exports[`ChoiceInput snapshots rationale does not render 1`] = `
<div
  className="corespring-checkbox"
>
  <div>
    <WithStyles(FeedbackTick)
      correctness="correct"
    />
    <div>
      <WithStyles(Component)
        control={
          <WithStyles(Component)
            checked={false}
            correctness="correct"
            onChange={[Function]}
          />
        }
        disabled={false}
        label="1. "
      />
      <Component
        className="label"
        onClick={[Function]}
        prompt="label"
        tagName="span"
      />
    </div>
  </div>
  <Feedback
    correctness="correct"
  />
</div>
`;

exports[`ChoiceInput snapshots rationale renders 1`] = `
<div
  className="corespring-checkbox"
>
  <div>
    <WithStyles(FeedbackTick)
      correctness="correct"
    />
    <div>
      <WithStyles(Component)
        control={
          <WithStyles(Component)
            checked={false}
            correctness="correct"
            onChange={[Function]}
          />
        }
        disabled={false}
        label="1. "
      />
      <Component
        className="label"
        onClick={[Function]}
        prompt="label"
        tagName="span"
      />
    </div>
  </div>
  <Component
    className="rationale"
    prompt="This is rationale"
  />
  <Feedback
    correctness="correct"
  />
</div>
`;
